### PR TITLE
Simplifying MCSync

### PR DESCRIPTION
**Pull request by RavenholmZombie**
**Changelog:**

- MCSync will now automatically create folders in directories specified by the user.
- Users can now just point MCSync to their `.minecraft` folder and MCSync will automatically create or use the required folders inside.
   - In addition, users can also opt to override this and use separate folders if desired.
- The MesaSuite Dashboard can now be resized.
- The "Sync in Progress" window has been given a minor UI update (see image below).

**Images**

_The new MCSync icon_

_Sync window showing new UI update_

### CODE REVIEW

File: MCSync/FancyUX.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        [Category("RZ's Fancy Buttons")]
+        public Color TextColor
+        {
+            get { return this.ForeColor; }
+            set { this.ForeColor = value; }
+        }
+
+        public int BorderSize1 { get => borderSize; set => borderSize = value; }
+        public int BorderRadius1 { get => borderRadius; set => borderRadius = value; }
+        public Color BorderColor1 { get => borderColor; set => borderColor = value; }
```
Comment (CSX8600) [inline MCSync/FancyUX.cs:71]: These can probably be removed - I don't see any references to them anywhere

File: MCSync/FancyUX.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+using System.Threading.Tasks;
+
+using System.Windows.Forms;
+using System.Drawing;
+using System.Drawing.Drawing2D;
+using System.ComponentModel;
+
+namespace MCSync
+{
+    public class FancyButton : Button
```
Comment (CSX8600) [inline MCSync/FancyUX.cs:14]: We should name the file the same as the class

File: MCSync/Syncer.cs [hunk 3]
```diff
@@ -72,6 +72,19 @@ private async System.Threading.Tasks.Task DoSync()
...
+                try
+                {
+                    Directory.CreateDirectory(modsDirectory);
+                    Directory.CreateDirectory(resourcePackDirectory);
+                    Directory.CreateDirectory(configFilesDirectory);
+                    Directory.CreateDirectory(oResourcesDirectory);
+                }
+                catch(IOException iox)
+                {
+                    Task.Errors.Add($"An unexpected error occurred during Sync: {iox.Message}");
```
Comment (CSX8600) [inline MCSync/Syncer.cs:85]: I don't think we even need to put this in a try/catch. The outer try/catch will do exactly this

File: MCSync/frmConfig.cs [hunk 4]
```diff
@@ -90,6 +62,11 @@ private void frmConfig_Load(object sender, EventArgs e)
 
             rbClient.Checked = syncMode == SyncMode.Client;
             rbServer.Checked = syncMode == SyncMode.Server;
+
+            overrideFoldersCheckBox.Checked = string.IsNullOrEmpty(configValues.GetOrDefault("minecraftDirectory").Cast<string>());
+            overrideFoldersCheckBox_CheckedChanged(this, EventArgs.Empty);
```
Comment (CSX8600) [inline MCSync/frmConfig.cs:67]: Pass in `overrideFoldersCheckBox` instead of `this` for the first parameter

File: MCSync/frmConfig.cs [hunk 5]
```diff
@@ -112,7 +89,16 @@ private void cmdModsWhitelist_Click(object sender, EventArgs e)
             whitelist.Text = "Mods Whitelist";
             whitelist.WhitelistName = "mods_whitelist";
             whitelist.lblIntro.Text = "Edit your Mods whitelist.";
-            whitelist.ShowDialog();
+
+            whitelist.TopLevel = false;
+            whitelist.ControlBox = false;
+            whitelist.Dock = DockStyle.Fill;
+            whitelist.FormBorderStyle = FormBorderStyle.None;
```
Comment (CSX8600) [inline MCSync/frmConfig.cs:96]: Should we actually just set these on the form itself rather than doing it here?

File: MCSync/frmConfig.cs [hunk 6]
```diff
@@ -121,7 +107,16 @@ private void cmdResourcePacksWhitelist_Click(object sender, EventArgs e)
             whitelist.Text = "Resource Packs Whitelist";
             whitelist.WhitelistName = "resourcepacks_whitelist";
             whitelist.lblIntro.Text = "Edit your Resource Packs whitelist.";
-            whitelist.ShowDialog();
+
+            whitelist.TopLevel = false;
+            whitelist.ControlBox = false;
+            whitelist.Dock = DockStyle.Fill;
+            whitelist.FormBorderStyle = FormBorderStyle.None;
```
Comment (CSX8600) [inline MCSync/frmConfig.cs:114]: Should we actually just set these on the form itself rather than doing it here?

File: MCSync/frmConfig.cs [hunk 7]
```diff
@@ -137,5 +132,135 @@ private void cmdBrowseOResources_Click(object sender, EventArgs e)
...
+                }
+
+                txtConfigDirectory.Text = txtMinecraftFolder.Text + "\\config";
+                txtOResourcesDirectory.Text = txtMinecraftFolder.Text + "\\oresources";
+            }
+        }
+
+        private void rbServer_CheckedChanged(object sender, EventArgs e)
+        {
+            if(rbServer.Checked)
```
Comment (CSX8600) [inline MCSync/frmConfig.cs:184]: We should probably only do this logic if `overrideFoldersCheckBox` isn't checked, much like how you did in the method above

File: MCSync/frmMain.cs [hunk 8]
```diff
@@ -14,95 +23,77 @@ public frmMain()
...
 
-        private void newBtnSync_MouseClick(object sender, MouseEventArgs e)
-        {
-            frmSync sync = new frmSync();
-            sync.Show();
-            Hide();
+            config.TopLevel = false;
+            config.ControlBox = false;
+            config.Dock = DockStyle.Fill;
+            config.FormBorderStyle = FormBorderStyle.None;
```
Comment (CSX8600) [inline MCSync/frmMain.cs:62]: Should these options be set on the form itself rather than manually here?

File: MCSync/frmMain.cs [hunk 9]
```diff
@@ -14,95 +23,77 @@ public frmMain()
...
-        private void btnHelp_MouseEnter(object sender, EventArgs e)
+        private void fButtonSync_Click(object sender, EventArgs e)
         {
-            lblDocumentation.Visible = true;
-        }
+            frmSync sync = new frmSync();
+            sync.TopLevel = false;
+            sync.ControlBox = false;
+            sync.Dock = DockStyle.Fill;
+            sync.FormBorderStyle = FormBorderStyle.None;
```
Comment (CSX8600) [inline MCSync/frmMain.cs:79]: Set on form itself?

File: MCSync/frmMain.cs [hunk 10]
```diff
@@ -14,95 +23,77 @@ public frmMain()
...
+            sync.Show();
         }
 
-        public void backgroundChooser()
+        private void pnlMainTimer_Tick(object sender, EventArgs e)
         {
-            int bg = new Random().Next(0, 6);
-            switch(bg)
+            var frmConfig = Application.OpenForms["frmConfig"];
+            var frmSync = Application.OpenForms["frmSync"];
```
Comment (CSX8600) [inline MCSync/frmMain.cs:88]: We should use concrete types when we know them (in this case, `Form`).

Also, wouldn't these potentially cause an error if we never opened either of these forms?

File: MCSync/frmSync.cs [hunk 11]
```diff
@@ -1,22 +1,46 @@
...
         private void frmSync_Load(object sender, EventArgs e)
         {
+            formFadeTimer.Start();
+            Dictionary<string, object> configValues = UserPreferences.Get().Sections.GetOrSetDefault("mcsync", () => new Dictionary<string, object>());
+            showBalloonTips = configValues.GetOrDefault("showBalloonTips", true).Cast<bool>();
+
             Syncer syncer = new Syncer();
             syncer.TaskAdded += Syncer_TaskAdded;
             syncer.SyncComplete += Syncer_SyncComplete;
+            ControlBox = false;
```
Comment (CSX8600) [inline MCSync/frmSync.cs:30]: Set on form directly using designer?

File: MCSync/frmSync.cs [hunk 12]
```diff
@@ -53,6 +109,8 @@ private void Syncer_TaskAdded(object sender, Task e)
                 rowsByTask[e] = row;
 
                 e.StatusUpdate += Task_StatusUpdate;
+                pbarOverall.Maximum = dgvTasks.Rows.Count;
+                ControlBox = false;
```
Comment (CSX8600) [inline MCSync/frmSync.cs:113]: ?

**Review by CSX8600 — COMMENTED** [review]
Just some basic comments worth looking over. Look 'em over, make changes if you want, and then resubmit when finished

File: MCSync/FancyUX.cs (hunk 1 above)
Comment (RavenholmZombie) [reply MCSync/FancyUX.cs:71]: Those variables are auto-generated and the class breaks without them.

File: MCSync/FancyUX.cs (hunk 2 above)
Comment (RavenholmZombie) [reply MCSync/FancyUX.cs]: Fixed

File: MCSync/Syncer.cs (hunk 3 above)
Comment (RavenholmZombie) [reply MCSync/Syncer.cs]: Fixed

File: MCSync/frmConfig.cs (hunk 5 above)
Comment (RavenholmZombie) [reply MCSync/frmConfig.cs]: Fixed

File: MCSync/frmConfig.cs (hunk 6 above)
Comment (RavenholmZombie) [reply MCSync/frmConfig.cs]: Fixed

File: MCSync/frmMain.cs (hunk 8 above)
Comment (RavenholmZombie) [reply MCSync/frmMain.cs]: Fixed

File: MCSync/frmConfig.cs (hunk 4 above)
Comment (RavenholmZombie) [reply MCSync/frmConfig.cs]: Fixed :3

File: MCSync/frmMain.cs (hunk 9 above)
Comment (RavenholmZombie) [reply MCSync/frmMain.cs]: Fixed

File: MCSync/frmMain.cs (hunk 10 above)
Comment (RavenholmZombie) [reply MCSync/frmMain.cs:82]: This code is crappy, but it's vital 🤣🤣🤣🤣

File: MCSync/frmSync.cs (hunk 12 above)
Comment (RavenholmZombie) [reply MCSync/frmSync.cs]: Fixed

File: MCSync/frmSync.cs (hunk 11 above)
Comment (RavenholmZombie) [reply MCSync/frmSync.cs]: Fixed

File: MCSync/frmConfig.cs (hunk 7 above)
Comment (RavenholmZombie) [reply MCSync/frmConfig.cs]: Fixed

File: MCSync/frmMain.cs (hunk 10 above)
Comment (RavenholmZombie) [reply MCSync/frmMain.cs]: Fixed

**Review by CSX8600 — APPROVED** [review]

**Pull request merged**